plugins/imfile/imfile.c

```diff
@@ -770,12 +853,91 @@ poll_tree(fs_edge_t *const chld)
+			issymlink = isSymlink(file, (char*)chld->path, &target);
+			if (issymlink == -1) { /* some serious problem with the file - skip */
+				continue;
+			}
+			/* need to add watch for symlink as well in case of target change */
+			const int is_file = (S_ISREG(fileInfo.st_mode) || issymlink);
+			DBGPRINTF("poll_tree:  found '%s', File: %d (config file: %d), symlink: %d\n",
+				file, is_file, chld->is_file, issymlink);
+			if (issymlink > 0) {
+				int rv;
+				size_t len = strlen((const char *)chld->path) + strlen(file) + 2;
+				char *slink = malloc(len);
+				char *ptr;
+				int freetarget = 1;
+				if (NULL == slink) {
+					LogError(errno, RS_RET_OUT_OF_MEMORY, "imfile: poll_tree: \
+							malloc failed");
+					free(target);
+					continue;
+				}
+				snprintf(slink, len, "%s/%s", chld->path, file);
+				/* Check if slink -> target is already in the hashtable. */
+				ptr = hashtable_search(slink_target, slink);
+				rv = 1;
+				if (NULL != ptr) { /* slink -> something is in the hashtable. */
+					if (0 != strcmp(ptr, target)) { /* something is not target. */
+						ptr = hashtable_remove(slink_target, slink);
+						free(ptr);
+						DBGPRINTF("imfile: poll_tree: \
+							hashtable_insert(slink, %s->%s)\n", slink, target);
+						/* Note: slink & target are passed in and consumed in the hashtable. */
+						rv = hashtable_insert(slink_target, slink, target);
+						freetarget = 0;
+					}
+				} else {
+					DBGPRINTF("imfile: poll_tree: \
+							hashtable_insert(slink, %s->%s)\n", slink, target);
+					/* Note: slink & target are passed in and consumed in the hashtable. */
+					rv = hashtable_insert(slink_target, slink, target);
+					freetarget = 0;
+				}
+				if(rv == 0) { /* out of memory */
+					LogError(errno, RS_RET_OUT_OF_MEMORY, "imfile: poll_tree: \
+							hastable_insert failed");
+					if (freetarget) {
+						free(target);
+					}
+					continue;
+				}
+				/* Check if target -> slink is already in the hashtable. */
+				ptr = hashtable_search(target_slink, target);
+				rv = 1;
+				if (NULL != ptr) { /* target -> something is in the hashtable. */
+					if (0 != strcmp(ptr, slink)) { /* something is not slink. */
+						ptr = hashtable_remove(target_slink, target);
+						free(ptr);
+						DBGPRINTF("imfile: poll_tree: \
+							hashtable_insert(target_slink, %s->%s)\n", target, slink);
+						rv = hashtable_insert(target_slink, strdup(target), strdup(slink));
+					}
+				} else {
+					DBGPRINTF("imfile: poll_tree: \
+							hashtable_insert(target_slink, %s->%s)\n", target, slink);
+					rv = hashtable_insert(target_slink, strdup(target), strdup(slink));
+				}
+				if(rv == 0) { /* out of memory */
+					LogError(errno, RS_RET_OUT_OF_MEMORY, "imfile: poll_tree: \
+							hastable_insert failed");
+					if (freetarget) {
+						free(target);
+					}
+					continue;
+				}
+				DBGPRINTF("imfile: %s -> %s added to symlink hash\n", slink, target);
```

Comment:
Do we need to call realloc to expand files.gl_pathv to allocate the space for target prior to the line 931 (as being done in the downstream version)?

Reply:
Yes, I will edit the PR once I finish first round of testing.